hibernate-core/src/main/java/org/hibernate/action/internal/EntityInsertAction.java

```diff
@@ -150,6 +153,168 @@ public void execute() throws HibernateException {
 
 		handleNaturalIdPostSaveNotifications( id );
 
+		EntityMetamodel entityMetamodel = getPersister().getEntityMetamodel();
+
+		if (entityMetamodel.getBytecodeEnhancementMetadata().isEnhancedForLazyLoading()) {
+			PersistentAttributeInterceptable interceptable = ((PersistentAttributeInterceptable) instance );
+			PersistentAttributeInterceptor interceptor = interceptable.$$_hibernate_getInterceptor();
+			final BytecodeEnhancementMetadata enhancementMetadata = entityMetamodel.getBytecodeEnhancementMetadata();
+			final LazyAttributesMetadata lazyAttributesMetadata = enhancementMetadata.getLazyAttributesMetadata();
+			if (interceptor == null) {
+				Type[] types = persister.getPropertyTypes();
+				final String[] propertyNames = persister.getPropertyNames();
+				final CascadeStyle[] cascadeStyles = persister.getPropertyCascadeStyles();
+				boolean needsInterceptor = false;
+				Set<String> initializedLazyFields = new HashSet<>();
+				for ( int i = 0; i < types.length; i++) {
+					if(!types[i].isAssociationType()) {
+						continue;
+					}
```

Comment:
Copying the [conversation from the original PR](https://github.com/hibernate/hibernate-orm/pull/4339#discussion_r742641548).

From @yrodiere:

> This effectively considers every `*ToOne` association as uninitialized, even if it has a non-null value. That doesn't seem like a good idea?

From @busitech:

> Yes, that is essential.  Just having a value doesn't mean Hibernate Search can traverse any association properties  during indexing unless interceptors are installed.  Assume the value was set by the client, and is "dead" from a perspective of indexing.  Imagine that the present object is only a small part of a huge multidimensional object graph, and that your search indexes depend on being able to traverse both up and down the hierarchical tree, using a proxy installed at every *ToOne and *ToMany property, otherwise your index will be incomplete and corrupt!

Reply:
It actually replaces every *ToOne non-proxy with a HibernateProxy (or injects the interceptor and marks the property unfetched) so that Hibernate Search can go through that node to the next @IndexedEmbedded association that it needs to load in order to reindex.  Every non-null non-proxy like a detached value is replaced with a proxy because all objects have relationships beyond them.

Reply:
The only time a property is marked initialized is if we find the interceptor already injected AND the object is initialized.  We assume this means the injector did the init, which means that the next level (up and down the hierarchical tree) can be traversed during indexing.  This scenario happens when server side code does an `em.find` and then assigns that entity to a property of another object which gets persisted.

Reply:
>  Every non-null non-proxy like a detached value is replaced with a proxy

Then that's, IMO, a different problem: you're no longer filling the gaps in associations in a newly inserted object, you're actually correcting wrong references in a graph; you're essentially doing half a merge during a persist.

Right now, applications are not supposed to assign detached values to properties. If all they have is a reference to a detached entity, they must retrieve an attached version of that entity first (e.g. through a merge, or by calling `session.getReference()` to retrieve a proxy), then assign that value to the property.

Knowing that neither a merge nor calling `session.getReference` involves a database call, this time I don't think one can argue this is not doable for performance reasons.

This new behavior you're proposing *could* make sense, but it seems like a different problem you're trying to fix, and I don't think you're fixing it completely. For example, detached entities could be assigned to already-persisted entities, for an update. When will *those* detached entities be replaced with attached proxies?

It also raises the question of cascades, which you saw and sidestepped by not applying this change on associations with any kind of cascade. But not planning for that could result in inconsistent behavior. We need to at least have an idea of the desired behavior when cascades are enabled, before we apply such a change.

If you want to discuss this further, I think the best way to actually make progress would be to talk to the Hibernate ORM team directly, rather than me: https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev/